### DISCUSSION TITLE
Contribute: Fetch regions and plans for DO, AWS, Vultr and Linode #327

### DISCUSSION THREAD

**Discussion by joglomedia** [vitodeploy/vito#338] [Ideas]
@saeedvaziry Hi there! Is anyone working on issue #327? If not, I am willing to take it on. Here is my current work in progress (local).

### Current WIP

**In progress:**
* [x]  AWS (done)
* [x]  Linode (done)
* [x]  DigitalOcean (done)
* [x]  Vultr (done)
* [x]  Hetzner (done)

**Proposed new server providers:**
* [x]  Amazon Lightsail (an affordable alternative to AWS EC2 for new users)
* [ ]  OVH
* [ ]  UpCloud

### Proposed Changes:

- Added `zone` to AWS, as each AWS zone offers different available instance types (plans).
- Images will be fetched directly from the API. From my testing results, each AWS region has different available images, so we can't use predefined image IDs.
- Since all server provider data is fetched from the API, the `config/serverproviders.php` file could be deprecated in the future.

### PoC / Preview:
I'll submit the code as soon as possible.

**Comment by saeedvaziry** [comment]
Hey @joglomedia,
I don't think anyone working on this at the moment.
PRs are welcome.
By zones for AWS, you mean the same regions right?

**Comment by joglomedia** [comment]
@saeedvaziry Yes, here is a preview about the AWS region & region's zone

https://www.awesomescreenshot.com/video/33426163?key=4834dc786b19f8eac95c8402f9042a01

I'll send PR soon

**Comment by saeedvaziry** [comment]
I get your point. Waiting for the PR.

Just a thought, Lets not bring the image selection into this. To deliver a stable software we need to focus on one and make sure that works. so lets hardcode the official images.

**Reply by joglomedia** [reply]
> I get your point. Waiting for the PR.
> 
> Just a thought, Lets not bring the image selection into this. To deliver a stable software we need to focus on one and make sure that works. so lets hardcode the official images.

Okay I understand, my first consideration is that the provided image may change or be deprecated in the future (such as what AWS does). Thus we need to actively check the pre-defined image whether it is still available or not.

If so, we must first determine which image will be used as the pre-defined image. I will try to list all the images available from each provider, then we determine which image will be selected as the default pre-defined image. 😃

**Reply by saeedvaziry** [reply]
We can select one of the images dynamically.

**Reply by joglomedia** [reply]
> We can select one of the images dynamically.

My latest previews dynamically fetch the image from the API. Currently only on AWS, resulting in multiple variations of images for each Ubuntu version.

The image fetching uses the selected Ubuntu version from `core.operating_systems` value as a filter. I'will try to fine tune the filter.

**Comment by saeedvaziry** [comment]
@joglomedia feel free to assign the issue to yourself 🚀

**Comment by saeedvaziry** [comment]
One last thing, I would appreciate one PR per provider instead of a big PR for the sake of easier review. 
So don't waste your time on a big PR.

**Comment by saeedvaziry** [comment]
@joglomedia As we are getting close to the end of November (releasing v2) I handled the necessary ones myself. feel free to raise a separate issue for Lightsail, OVH and UpCloud
